committing large file to HTTPRepository throws RejectedExecutionException

Hi,

I want to write large RDF turtle files (up to 100 MB) via RDF4J HTTP client (version 2.2.2) to a java native store in RDF4J Server (version 2.2.2) via this java code:

```java
ValueFactory valueFactory = SimpleValueFactory.getInstance();
final IRI iri = valueFactory.createIRI("urn:test_iri");
final IRI context = valueFactory.createIRI("urn:test_context");
Repository repository = new HTTPRepository("http://localhost:8080/rdf4j-server/", "test");
repository.initialize();
RepositoryConnection tripleStoreConnection = null;
try {
	tripleStoreConnection = repository.getConnection();
	tripleStoreConnection.begin();
        tripleStoreConnection.clear(context);
	tripleStoreConnection.add(file, iri, RDFFormat.TURTLE, context);
	tripleStoreConnection.commit();
} catch (final Exception e) {
	// print exception
} finally {
	try {
		if (tripleStoreConnection != null) {
			tripleStoreConnection.close();
		}
	} catch (final Exception e) {
		// print exception
	}
}
```

Client and server are running in the same tomcat instance on a CentOS7 VM.

Unfortunately in most cases (not always) the server throws a RejectedExecutionException (see below) during commit phase after up to 20 minutes. Curiously the commit works after several retries (the bigger the file the more retries are necessary). The last commit needs only a few seconds. During commit phase the number of disk writes per second is very high (up to 1300 writes per second) but the throughput per write is very small (up to 6 KB per write).

Does anybody has an explanation for this strange behavior? Is there an alternative way for fast and robust writing large RDF files to a HTTP repository via RDF4J?

Thanks in advance.

```
[INFO ] 2017-08-03 15:14:33,665 [repositories/test] POST transaction start
[INFO ] 2017-08-03 15:14:33,720 [repositories/test] transaction started
[INFO ] 2017-08-03 15:14:33,819 [repositories/test] PUT txn operation
[INFO ] 2017-08-03 15:14:33,996 [repositories/test] PUT txn operation request finished.
[INFO ] 2017-08-03 15:14:34,775 [repositories/test] PUT txn operation
[INFO ] 2017-08-03 15:26:09,885 [repositories/test] PUT txn operation request finished.
[WARN ] 2017-08-03 15:26:09,969 [rdf4j-transaction-0] Rolling back transaction due to connection close
[WARN ] 2017-08-03 15:26:09,970 [repositories/test] deregistered expired transaction cbc6cedf-6099-4ceb-8cdd-f9d8744a2dc8
[INFO ] 2017-08-03 15:26:09,972 [repositories/test] PUT txn operation
[ERROR] 2017-08-03 15:26:09,978 [repositories/test] Error while handling request (500)
org.eclipse.rdf4j.http.server.ServerHTTPException: Transaction handling error: Task java.util.concurrent.FutureTask@37727308 rejected from java.util.concurrent.ThreadPoolExecutor@69baa56d[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 5]
	at org.eclipse.rdf4j.http.server.repository.transaction.TransactionController.processModificationOperation(TransactionController.java:277)
	at org.eclipse.rdf4j.http.server.repository.transaction.TransactionController.handleRequestInternal(TransactionController.java:182)
	at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:147)
	at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:50)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:959)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:893)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doPut(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:653)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:731)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at com.github.ziplet.filter.compression.CompressingFilter.doFilter(CompressingFilter.java:300)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:505)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:169)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
	at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:956)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:436)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1078)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:625)
	at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:318)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.FutureTask@37727308 rejected from java.util.concurrent.ThreadPoolExecutor@69baa56d[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 5]
	at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2063)
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:830)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1379)
	at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:134)
	at java.util.concurrent.Executors$DelegatedExecutorService.submit(Executors.java:681)
	at org.eclipse.rdf4j.http.server.repository.transaction.Transaction.submit(Transaction.java:502)
	at org.eclipse.rdf4j.http.server.repository.transaction.Transaction.commit(Transaction.java:168)
	at org.eclipse.rdf4j.http.server.repository.transaction.TransactionController.processModificationOperation(TransactionController.java:258)
	... 33 common frames omitted
[ERROR] 2017-08-03 15:26:10,093 [repositories/test] Error while handling request
org.eclipse.rdf4j.repository.RepositoryException: transaction with id cbc6cedf-6099-4ceb-8cdd-f9d8744a2dc8 not registered.
	at org.eclipse.rdf4j.http.server.repository.transaction.ActiveTransactionRegistry.getTransaction(ActiveTransactionRegistry.java:139)
	at org.eclipse.rdf4j.http.server.repository.transaction.TransactionController.handleRequestInternal(TransactionController.java:113)
	at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:147)
	at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:50)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:959)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:893)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:970)
	at org.springframework.web.servlet.FrameworkServlet.doDelete(FrameworkServlet.java:894)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:656)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:846)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:731)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at com.github.ziplet.filter.compression.CompressingFilter.doFilter(CompressingFilter.java:300)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:505)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:169)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
	at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:956)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:436)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1078)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:625)
	at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:316)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/http/client/src/main/java/org/eclipse/rdf4j/http/client/RDF4JProtocolSession.java
+++ b/http/client/src/main/java/org/eclipse/rdf4j/http/client/RDF4JProtocolSession.java
@@ -21,7 +21,9 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
-import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
@@ -76,6 +78,8 @@ import org.eclipse.rdf4j.rio.RDFHandler;
 import org.eclipse.rdf4j.rio.RDFHandlerException;
 import org.eclipse.rdf4j.rio.RDFParseException;
 import org.eclipse.rdf4j.rio.helpers.BasicParserSettings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * A {@link SPARQLProtocolSession} subclass which extends the standard SPARQL 1.1 Protocol with additional
@@ -87,12 +91,27 @@ import org.eclipse.rdf4j.rio.helpers.BasicParserSettings;
  */
 public class RDF4JProtocolSession extends SPARQLProtocolSession {
 
+	/**
+	 * How long the client should wait before sending another PING to the server
+	 */
+	private static final long PINGDELAY = TimeUnit.MILLISECONDS.convert(Protocol.DEFAULT_TIMEOUT,
+			TimeUnit.SECONDS) / 2;
+
+	private final Logger logger = LoggerFactory.getLogger(RDF4JProtocolSession.class);
+
 	private String serverURL;
 
 	private String transactionURL;
 
-	public RDF4JProtocolSession(HttpClient client, ExecutorService executor) {
+	private ScheduledExecutorService executor;
+
+	private ScheduledFuture<?> ping;
+
+	private long pingDelay = PINGDELAY;
+
+	public RDF4JProtocolSession(HttpClient client, ScheduledExecutorService executor) {
 		super(client, executor);
+		this.executor = executor;
 
 		// we want to preserve bnode ids to allow Sesame API methods to match
 		// blank nodes.
@@ -102,6 +121,16 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 		// most performant
 		setPreferredTupleQueryResultFormat(TupleQueryResultFormat.BINARY);
 		setPreferredRDFFormat(RDFFormat.BINARY);
+		try {
+			final String configuredValue = System.getProperty(Protocol.CACHE_TIMEOUT_PROPERTY);
+			if (configuredValue != null) {
+				int timeout = Integer.parseInt(configuredValue);
+				pingDelay = TimeUnit.MILLISECONDS.convert(Math.max(timeout, 1), TimeUnit.SECONDS) / 2;
+			}
+		}
+		catch (Exception e) {
+			logger.warn("Could not read integer value of system property {}", Protocol.CACHE_TIMEOUT_PROPERTY);
+		}
 	}
 
 	public void setServerURL(String serverURL) {
@@ -147,6 +176,10 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 	@Override
 	public String getUpdateURL() {
 		return Protocol.getStatementsLocation(getQueryURL());
+	}
+
+	private synchronized String getTransactionURL() {
+		return transactionURL;
 	}
 
 	/*-----------------*
@@ -223,6 +256,7 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 		checkRepositoryURL();
 
 		try {
+			String transactionURL = getTransactionURL();
 			final boolean useTransaction = transactionURL != null;
 
 			String baseLocation = useTransaction ? appendAction(transactionURL, Action.SIZE)
@@ -239,6 +273,7 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 
 			try {
 				String response = EntityUtils.toString(executeOK(method).getEntity());
+				pingTransaction();
 
 				try {
 					return Long.parseLong(response);
@@ -463,6 +498,7 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 		checkRepositoryURL();
 
 		try {
+			String transactionURL = getTransactionURL();
 			final boolean useTransaction = transactionURL != null;
 
 			String baseLocation = useTransaction ? transactionURL
@@ -502,6 +538,7 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 		catch (URISyntaxException e) {
 			throw new AssertionError(e);
 		}
+		pingTransaction();
 	}
 
 	public synchronized void beginTransaction(IsolationLevel isolationLevel)
@@ -533,6 +570,8 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 					transactionURL = response.getFirstHeader("Location").getValue();
 					if (transactionURL == null) {
 						throw new RepositoryException("no valid transaction ID received in server response.");
+					} else {
+						pingTransaction();
 					}
 				}
 				else {
@@ -569,6 +608,9 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 				if (code == HttpURLConnection.HTTP_OK) {
 					// we're done.
 					transactionURL = null;
+					if (ping != null) {
+						ping.cancel(false);
+					}
 				}
 				else {
 					throw new RepositoryException("unable to commit transaction. HTTP error code " + code);
@@ -608,6 +650,9 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 				if (code == HttpURLConnection.HTTP_NO_CONTENT) {
 					// we're done.
 					transactionURL = null;
+					if (ping != null) {
+						ping.cancel(false);
+					}
 				}
 				else {
 					throw new RepositoryException("unable to rollback transaction. HTTP error code " + code);
@@ -620,6 +665,45 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 		finally {
 			method.reset();
 		}
+	}
+
+	private synchronized void pingTransaction() {
+		if (transactionURL == null) {
+			return;
+		}
+		if (ping != null) {
+			ping.cancel(false);
+		}
+		if (pingDelay > 0) {
+			ping = executor.schedule(() -> {
+				executeTransactionPing();
+			}, pingDelay, TimeUnit.MILLISECONDS);
+		}
+	}
+
+	void executeTransactionPing() {
+		String transactionURL = getTransactionURL();
+		if (transactionURL == null) {
+			return; // transaction has already been closed
+		}
+		HttpPost method = null;
+		try {
+			URIBuilder url = new URIBuilder(transactionURL);
+			url.addParameter(Protocol.ACTION_PARAM_NAME, Action.PING.toString());
+			method = new HttpPost(url.build());
+			String text = EntityUtils.toString(executeOK(method).getEntity());
+			long timeout = Long.parseLong(text);
+			// clients should ping before server timeouts transaction
+			long nextPingDelay = timeout /2;
+			synchronized (this) {
+				if (pingDelay != nextPingDelay) {
+					pingDelay = nextPingDelay;
+				}
+			}
+		} catch (Exception e) {
+			logger.warn("Failed to ping transaction", e.toString());
+		}
+		pingTransaction(); // reschedule
 	}
 
 	/**
@@ -745,6 +829,7 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 			boolean includeInferred, int maxQueryTime, Binding... bindings)
 	{
 		RequestBuilder builder = null;
+		String transactionURL = getTransactionURL();
 		if (transactionURL != null) {
 			builder = RequestBuilder.put(transactionURL);
 			builder.setHeader("Content-Type", Protocol.SPARQL_QUERY_MIME_TYPE + "; charset=utf-8");
@@ -757,6 +842,7 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 			// in a PUT request, we carry the actual query string as the entity
 			// body rather than a parameter.
 			builder.setEntity(new StringEntity(query, UTF8));
+			pingTransaction();
 		}
 		else {
 			builder = RequestBuilder.post(getQueryURL());
@@ -778,6 +864,7 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 			boolean includeInferred, int maxExecutionTime, Binding... bindings)
 	{
 		RequestBuilder builder = null;
+		String transactionURL = getTransactionURL();
 		if (transactionURL != null) {
 			builder = RequestBuilder.put(transactionURL);
 			builder.addHeader("Content-Type", Protocol.SPARQL_UPDATE_MIME_TYPE + "; charset=utf-8");
@@ -790,6 +877,7 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 			// in a PUT request, we carry the only actual update string as the
 			// request body - the rest is sent as request parameters
 			builder.setEntity(new StringEntity(update, UTF8));
+			pingTransaction();
 		}
 		else {
 			builder = RequestBuilder.post(getUpdateURL());
@@ -869,6 +957,7 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 
 		checkRepositoryURL();
 
+		String transactionURL = getTransactionURL();
 		boolean useTransaction = transactionURL != null;
 
 		try {
@@ -933,6 +1022,7 @@ public class RDF4JProtocolSession extends SPARQLProtocolSession {
 		catch (URISyntaxException e) {
 			throw new AssertionError(e);
 		}
+		pingTransaction();
 	}
 
 	@Override

--- a/http/client/src/main/java/org/eclipse/rdf4j/http/client/SesameClientImpl.java
+++ b/http/client/src/main/java/org/eclipse/rdf4j/http/client/SesameClientImpl.java
@@ -7,7 +7,7 @@
  *******************************************************************************/
 package org.eclipse.rdf4j.http.client;
 
-import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
 
 import org.apache.http.impl.client.CloseableHttpClient;
 
@@ -30,7 +30,7 @@ public class SesameClientImpl extends SharedHttpClientSessionManager {
 	 * @deprecated use {@link SharedHttpClientSessionManager} instead.
 	 */
 	@Deprecated
-	public SesameClientImpl(CloseableHttpClient dependentClient, ExecutorService dependentExecutorService) {
+	public SesameClientImpl(CloseableHttpClient dependentClient, ScheduledExecutorService dependentExecutorService) {
 		super(dependentClient, dependentExecutorService);
 	}
 

--- a/http/client/src/main/java/org/eclipse/rdf4j/http/client/SesameSession.java
+++ b/http/client/src/main/java/org/eclipse/rdf4j/http/client/SesameSession.java
@@ -7,7 +7,7 @@
  *******************************************************************************/
 package org.eclipse.rdf4j.http.client;
 
-import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
 
 import org.apache.http.client.HttpClient;
 
@@ -22,7 +22,7 @@ public class SesameSession extends RDF4JProtocolSession {
 	 * @deprecated use {@link RDF4JProtocolSession} instead.
 	 */
 	@Deprecated
-	public SesameSession(HttpClient client, ExecutorService executor) {
+	public SesameSession(HttpClient client, ScheduledExecutorService executor) {
 		super(client, executor);
 	}
 

--- a/http/client/src/main/java/org/eclipse/rdf4j/http/client/SharedHttpClientSessionManager.java
+++ b/http/client/src/main/java/org/eclipse/rdf4j/http/client/SharedHttpClientSessionManager.java
@@ -10,8 +10,8 @@ package org.eclipse.rdf4j.http.client;
 import java.util.Map;
 import java.util.Objects;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ThreadFactory;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicLong;
@@ -41,7 +41,7 @@ public class SharedHttpClientSessionManager implements HttpClientSessionManager,
 	/** dependent life cycle */
 	private volatile CloseableHttpClient dependentClient;
 
-	private final ExecutorService executor;
+	private final ScheduledExecutorService executor;
 
 	/**
 	 * Optional {@link HttpClientBuilder} to create the inner {@link #httpClient} (if not provided externally)
@@ -56,7 +56,7 @@ public class SharedHttpClientSessionManager implements HttpClientSessionManager,
 
 	public SharedHttpClientSessionManager() {
 		final ThreadFactory backingThreadFactory = Executors.defaultThreadFactory();
-		this.executor = Executors.newCachedThreadPool(new ThreadFactory() {
+		this.executor = Executors.newScheduledThreadPool(0, new ThreadFactory() {
 
 			public Thread newThread(Runnable runnable) {
 				Thread thread = backingThreadFactory.newThread(runnable);
@@ -68,7 +68,7 @@ public class SharedHttpClientSessionManager implements HttpClientSessionManager,
 	}
 
 	public SharedHttpClientSessionManager(CloseableHttpClient dependentClient,
-			ExecutorService dependentExecutorService)
+			ScheduledExecutorService dependentExecutorService)
 	{
 		this.httpClient = this.dependentClient = Objects.requireNonNull(dependentClient,
 				"HTTP client was null");

--- a/http/protocol/src/main/java/org/eclipse/rdf4j/http/protocol/Protocol.java
+++ b/http/protocol/src/main/java/org/eclipse/rdf4j/http/protocol/Protocol.java
@@ -35,11 +35,44 @@ public abstract class Protocol {
 		QUERY,
 		/** SPARQL Update */
 		UPDATE,
+		/** Keep alive ping @since 2.3 */
+		PING,
 		/** commit */
 		COMMIT,
 		/** rollback */
 		ROLLBACK;
 	}
+
+	/**
+	 * Use an enum to allow ActiveTransactionRegistry enum to access the static values. This is necessary
+	 * because enum are initialize before static fields.
+	 */
+	@Deprecated
+	public static enum TIMEOUT {
+		CACHE;
+
+		/**
+		 * Configurable system property {@code rdf4j.server.txn.registry.timeout} for specifying the
+		 * transaction cache timeout (in seconds).
+		 */
+		public static final String CACHE_PROPERTY = "rdf4j.server.txn.registry.timeout";
+
+		/**
+		 * Default timeout setting for transaction cache entries (in seconds).
+		 */
+		public final static int DEFAULT = 60;
+	}
+
+	/**
+	 * Configurable system property {@code rdf4j.server.txn.registry.timeout} for specifying the transaction
+	 * cache timeout (in seconds).
+	 */
+	public static final String CACHE_TIMEOUT_PROPERTY = Protocol.TIMEOUT.CACHE_PROPERTY;
+
+	/**
+	 * Default timeout setting for transaction cache entries (in seconds).
+	 */
+	public final static int DEFAULT_TIMEOUT = Protocol.TIMEOUT.DEFAULT;
 
 	/**
 	 * Protocol version.

